### ISSUE TITLE
Is there any (rough) timeframe on MSBUILD15, NuGet4 and Visual Studio 2017

### ISSUE THREAD

**Issue by am11** [projectkudu/kudu#2376]
### Our Scenario:

I have upgraded one of our hybrid solution from XPROJ+CSPROJ based projects to VS2017 compatible (new-CSPROJ + old-CSPROJ) based infrastructure. That requires MSBuild v15.0, Nuget v4.0 and some Visual Studio 2017 to be present on the build machine. As we have auto deployment from source control, we are blocked to exercise it in production due to these missing tools in App Service instances and continuing to rely on outdated dotnet-cli v1.0.0-preview2-xxx (x64).

### Background:

Usually, people are either using solutions (SLN) comprised of "all" .NET Core projects (e.g. new CSPROJ) or "all" .NET Framework (old CSPROJ) based projects. The former is built either by dotnet-cli or MSBuild 15. Whereas, the latter one is built only by `Visual Studio <many versions>` (MSBuild 14, 15 etc.).

Then there are Hybrid or Mixed solutions, which use either both old and new CSPROJs added to SLN or all new CSPROJs that are targeting .NET Framework 4.6 for example. These kind of solutions are only buildable by Visual Studio 2017 (MSBuild 15 and NuGet4.0).

In a nutshell; the Roslyn project system, that is based on the Visual Studio 2017's Common Project System (CPS), is built on top of the .NET SDK that supports multiple Target Framework Monikers (TFM). Which means in the new kind of CSPROJ file, we can specify the `<TargetFrameworks>` (plural) member with multiple values (not limited to those of .NET Standard and .NET Core, but also) Xamarin and Full .NET Framework etc. as well. So the use-cases can be beyond the hybrid SLN.

### Problem:

`dotnet-cli` commands (`dotnet-restore`, `dotnet-build` etc.) only function on .NET Core TFMs (netcoreapp1.0 netstandard1.6 etc.), but not Full Framework monikers net45x, net46x.

The only toolset that support the RPS->CPS or Hybrid / Mixed solution workflow is: MSBuild15, NuGet4.0 and potentially some other components which gets installed with VS2017.

### Question:

Can we get any estimated timeframe (quarter / year) by which these tools would be available in Azure App Service instances?

Thanks for your continuous support! 🎉

**Comment by ctaggart** [comment]
Well explained! I'd really like to see MSBuild15 available too. #2350 Add "Build Tools for Visual Studio 2017" to Azure App Service asks for this.

**Comment by ebashmakov** [comment]
I'm having the same problem. We need msbuild15 in the App Services environment in order to smoothly migrate to the new csproj format and ASP.NET Core.

/cc @davidebbo @watashiSHUN

**Comment by watashiSHUN** [comment]
@am11 @ebashmakov @ctaggart 
we actually have msbuild15.3 preview in Azure, but it is not turned on by default.
you can see its location in kudu console:
```
D:\home>set msbuild
MSBUILD_15_DIR=D:\Program Files (x86)\MSBuild-15.3-preview\MSBuild\15.0\Bin
MSBUILD_PATH=D:\Program Files (x86)\MSBuild\14.0\Bin\MSBuild.exe
```
to use it you will need to have a custom deployment script with at least these two modification:
1. pass the msbuild15.3 directory to nuget (nuget4 has dependency on it):
`nuget.exe restore "{your .sln file path}" -MSBuildPath "%MSBUILD_15_DIR%"`
2. use msbuild15.3 to build and publish:
`"%MSBUILD_15_DIR%\MSBuild.exe {your build arguments}"`

[here](https://github.com/watashiSHUN/functionAppGitDeployment/commit/54475515da0ca8f6479020b24cb73bf7493f64fe) is an example

**Comment by ebashmakov** [comment]
@watashiSHUN thanks! It works for me.

**Comment by am11** [comment]
Thanks a bunch for making it happen @watashiSHUN! 👍 

As a workaround, I created a csproj file in new format, which -- by the virtue of GLOB pattern -- includes everything we needed from git submodules and referenced that in our solution instead. However, with GLOB, I hit this Solution Explorer blind spot issue dotnet/project-system#1970 (which is fixed in 15.3 project-system preview).

Nonetheless, it's a great news that we can have legacy project system support on Kudu! :)

Since we have an optable solution available in Azure now (and hoping on everyone's behalf that we would have this enabled by default once 15.3 toolchain ships out), I am closing this issue.
